transport-classes-io_uring/src/main/java/io/netty/channel/uring/IoUringIoHandler.java

```diff
@@ -221,11 +228,32 @@ private void submitEventFdRead() {
                 eventfd.intValue(), eventfdReadBuf, 0, 8, udata);
     }
 
-    private void submitTimeout(IoExecutorContext context) {
-        long delayNanos = context.delayNanos(System.nanoTime());
-        long udata = UserData.encode(RINGFD_ID, Native.IORING_OP_TIMEOUT, (short) 0);
+    private int submitAndWaitWithTimeout(SubmissionQueue submissionQueue,
+                                         boolean linkTimeout, long timeoutNanoSeconds) {
+        if (timeoutNanoSeconds != -1) {
+            long udata = UserData.encode(RINGFD_ID,
+                    linkTimeout ? Native.IORING_OP_LINK_TIMEOUT : Native.IORING_OP_TIMEOUT, (short) 0);
+            // We use the same timespec pointer for all add*Timeout operations. This only works because we call
+            // submit directly after it. This ensures the submitted timeout is considered "stable" and so can be reused.
+            long seconds, nanoSeconds;
+            if (timeoutNanoSeconds == 0) {
+                seconds = 0;
+                nanoSeconds = 0;
+            } else {
+                seconds = (int) min(timeoutNanoSeconds / 1000000000L, Integer.MAX_VALUE);
+                nanoSeconds = (int) max(timeoutNanoSeconds - seconds * 1000000000L, 0);
+            }
 
-        ringBuffer.ioUringSubmissionQueue().addTimeout(delayNanos, udata);
+            PlatformDependent.putLong(timeoutMemoryAddress + KERNEL_TIMESPEC_TV_SEC_FIELD, seconds);
+            PlatformDependent.putLong(timeoutMemoryAddress + KERNEL_TIMESPEC_TV_NSEC_FIELD, nanoSeconds);
+
+            if (linkTimeout) {
+                submissionQueue.addLinkTimeout(timeoutMemoryAddress, udata);
```

Comment:
I was still happy to have the sub q to NOT have the address as part of the API since you can compose as you wish the timespec type - but it's not a big deal actually

Reply:
@franz1981 I think its fine as we also take the address for example in `read` etc.